### DISCUSSION TITLE
ERROR: permission denied for table confirmation (SQLSTATE 42501) in custom auth flow

### DISCUSSION THREAD

**Discussion by Asaraspo** [supabase/supabase#15692] [Questions] [answered]
I am currently trying to implement a custom auth flow for my react native app. 

My auth flow should look like this for the sign up process:

SignUp -> navigates to an email confirmation screen -> awaits for email confirmation -> automatically navigates to the next screen upon confirmation

I am using realtime to check for updates to the confirmation status. Since the `auth.users` table cannot be used for realtime, I implemented a trigger and a function, which are updating a `public.confirmation` table. The `public.confirmation` table is used for realtime.

Whenever a user clicks on the link to confirm its email, I am getting an error with code 500. From the logs I have extraced the following error message `ERROR: permission denied for table confirmation (SQLSTATE 42501)`.

My `public.confirmation` table has no RLS activated. 

Also here is my trigger and my function which are used to update the email confirmation table.
```sql
CREATE OR REPLACE FUNCTION public.update_confirmation_email_status()
RETURNS trigger
LANGUAGE plpgsql
SET search_path = public, auth
AS $function$
BEGIN
 IF NEW.email_confirmed_at IS NOT NULL THEN
   INSERT INTO public.confirmation (id, email_confirmed)
   VALUES (NEW.id, TRUE);
 END IF;
 RETURN NEW;
END;
$function$;


CREATE TRIGGER update_confirmation_trigger
AFTER UPDATE ON auth.users
FOR EACH ROW
WHEN (OLD.email_confirmed_at IS DISTINCT FROM NEW.email_confirmed_at)
EXECUTE FUNCTION public.update_confirmation_email_status();
```

And also my table definition:
```sql
create table
  public.confirmation (
    id uuid not null,
    email_confirmed boolean not null default false,
    constraint confirmation_pkey primary key (id),
    constraint confirmation_id_fkey foreign key (id) references auth.users (id) on delete cascade
  ) tablespace pg_default;
```
Do you have any idea why the error is occuring?

Thanks in advance.

**Comment by GaryAustin1** [comment, marked as answer]
42501 is a grant/permission error not RLS policy error.  The issue is the auth role used by gotrue does not have access to the public schema.  Make your function `security definer` type and it should work.

**Reply by Asaraspo** [reply]
Thanks alot.
